native/common/include/jp_match.h

```diff
@@ -14,20 +14,43 @@ class JPMatch
 		_implicit = 2,
 		_exact = 3
 	} ;
+
+public:
+	JPMatch();
+	JPMatch(JPJavaFrame *frame, PyObject *object);
+
+	JPContext *getContext()
+	{
+		if (frame == NULL)
+			return NULL;
+		return frame->getContext();
+	}
+
+	/**
+	 * Get the Java slot associated with the Python object.
+	 *
+	 * Thus uses caching.
+	 *
+	 * @return the Java slot or 0 if not available.
+	 */
+	JPValue *getJavaSlot();
+
+	jvalue convert();
+
 public:
 	JPMatch::Type type;
-	JPConversion* conversion;
+	JPConversion *conversion;
+	JPJavaFrame *frame;
+	PyObject *object;
+	JPValue *slot;
+	void *closure;
```

Comment:
This is the heart of the PR as we are going to push a bunch of the brains back into the match.  

The frame and the object were basically being passed at every call when they were static to the match after creation.   
In addition every method was calling on the slot once in the match and then again in the convert method.  When we were doing a large overload we would get 5 or more times.  Thus caching it once is a performance issue.

Last the match often has something that needs to be passed to the convert (a class, proxy, or type) so we add a user defined pointer as the closure.